### ISSUE TITLE
New Check: instanceof should be named or unnamed

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#15179]
With the introduction of unnamed variables, it is possible to now write some old syntax with the presence of the unnamed variable and it basically have the same effect.

https://github.com/checkstyle/checkstyle/issues/14942 's best practice basically encourages using unnamed variables. We should have a check to enforce unnamed variables in the instances below to make it obvious to users that a variable is here, but just isn't being used at the moment. If something changes, they can change it to make it named.

````
import java.util.LinkedHashMap;

public class TestClass {
    void method() {
        Object m = "";

        if (m instanceof LinkedHashMap<?, ?>) { // expected violation
        }
        if (m instanceof LinkedHashMap<?, ?> _) { // expected fix
        }
        if (m instanceof LinkedHashMap<?, ?> named) { // acceptable fix if variable is used
        }
    }
}
````

This check would only be for higher Java versions.

**Comment by mahfouz72** [comment]
Some examples:

```Java

public class Test {

    void test(Object o) {
        if (o instanceof String) { // violation, pattern matching should declare a pattern variable
            String s = (String) o;
            System.out.println(s);
        }
        
        // after fix
        if (o instanceof String s) {
            System.out.println(s);
        }
    }
    
     void test2(Object o) {
        if (o instanceof String) { // violation, pattern matching should declare a pattern variable
            System.out.println("This is a String");
        }
        // after fix
        if (o instanceof String _) {
            System.out.println("This is a String");
        }
        if (o instanceof String s) { // this will be accepted also, the usage is not a concern for this check
            System.out.println("This is a String");
        }
    }
}
```

**Comment by nrmancuso** [comment]
After thinking on this check a bit, I believe that it is better to wait for some time to pass and this new language feature to become more widely used. Let's place this into "discussion ongoing" for now.

In some cases, like [MissingNullCaseInSwitch](https://github.com/checkstyle/checkstyle/pull/15113), it is clear that this new language feature targets a specific problem, and the purpose of this feature is unambiguous.

For unnamed variables, their usefulness is readily apparent when we MUST declare some variable, but don't need to.

Example:

```java
        int count = 0;
        for (int _ : ints) {
            count++;
        }
```

However, I am not completely sold on the usefulness of the following:
```java
        // would be a violation
        if (o instanceof String) {
            System.out.println("hello");
        }
        
        // makes this check happy
        if (o instanceof String _) {
            System.out.println("hello");
        }
```


I think it is easy enough to create some MatchXPath expression to force all `instanceof` usages to use some pattern variable for the time being:
```
➜  src cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="MatchXpath">
            <property name="query" value="//LITERAL_INSTANCEOF[not(./PATTERN_VARIABLE_DEF)]"/>
            <message key="matchxpath.match"
                     value="instanceof usage should declare a pattern or unnamed variable."/>
        </module>
    </module>
</module>

➜  src cat Test.java 
public class Test {
    void m(Object o) {
        if (o instanceof String) {
            System.out.println("hello");
        }

        if (o instanceof String _) {
            System.out.println("hello");
        }
    }
}

➜  src javac --enable-preview --release 21 Test.java
Note: Test.java uses preview features of Java SE 21.
Note: Recompile with -Xlint:preview for details.

➜  src java -jar checkstyle-10.17.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:3:15: instanceof usage should declare a pattern or unnamed variable. [MatchXpath]
Audit done.
Checkstyle ends with 1 errors.
```

**Comment by rnveach** [comment]
Byte code differences:
````
if (obj instanceof Number _) {}

       5: aload_1
       6: instanceof    #22                 // class java/lang/Number
       9: ifeq          22
      12: aload_1
      13: pop
````
````
if (obj instanceof Number) {

      22: aload_1
      23: instanceof    #22                 // class java/lang/Number
      26: ifeq          37
````

For the compiler I am using, `_` does have an unnecessary load and pop. Hopefully it will be fixed in future JREs, but the code is virtually the same. I would think the JIT would see these codes as unnecessary and end up removing them.

**Comment by nrmancuso** [comment]
While this seems like a reasonable check, I am removing this from the Java 21 Language Features project, and we should not pursue it until we have some demand from the community, IMO

**Comment by romani** [comment]
I can not find any reason to have extra characters in IF expression:
```
if (m instanceof Map<?, ?>) { // expected violation
}
if (m instanceof Map<?, ?> _) { // expected ok
}
```

Let's wait for explicit-lovers ( feedback from community)
